src/ide/tgui.ts

```diff
@@ -1146,14 +1164,329 @@ export let tgui = (function() {
+	window.addEventListener("resize", centerAllModalDialogs);
+
+	function nodeListContains(nodeList, element)
+	{
+		for (let i = 0; i < nodeList.length; i++) {
+			let item = nodeList[i];
+			if (item === element) return true;
+		}
+		return false;
+	}
+
+	// Create a modal dialog. Similar to createPanel
+	// The description object has the following fields:
+	// - title:             text in the title bar
+	// - scalesize:         [width, heigth] scaled size of the dialog
+	// - minsize:           [width, height] minimum size of the dialog, 
+	//                      when the whole viewport is smaller, the viewport size is used
+	// - contentstyle:      object to add/override some styles to/of the content element
+	// - onClose:           callback function() that is called when the dialog is closed by the user
+	// - buttons:           list of strings like ["Okay", "Cancel"], each button is connected to
+	//                      an eventhandler that is named on<Button>, if it is not available,
+	//                      onClose is used instead, if this is not given, there is no button bar
+	//                      at the bottom.
+	// - default_button     the name of the button in buttons, that is defaulted; it gets highlighted in blue
+	// - enter_confirms     boolean flag, if pressing [Enter] is like clicking on the default button
+	//
+	// those properties are carried over to the returned object
+	// and the following fields are contained in the returned object:
+	// - content:           a DOM element, that represents the content of the dialog
+	// - dom:               a DOM element, that represents the whole dialog
+	// - button_doms:       dictionary of DOM elements, that represents the buttons in the button bar
+	// - and others, mainly the titlebar components
+	//
+	// TODO: Support undecorated elements via a boolean property `decorated`, in this case
+	//       control.dom might be the same as control.content.
+	module.createModal = function(description)
+	{
+		let control = Object.assign({}, description);
+		// handle default fields
+		if (! control.hasOwnProperty("onClose")) control.onClose = function() { };
+		
+		// create dialog	
+		let dialog = tgui.createElement({
+			"type": "div",
+			"className": "tgui tgui-modal",
+			"style": {"background": "#eee", "overflow": "hidden", "display": "block", "zIndex": 100},
+		});
+		control.dom = dialog;
+		
+		dialog.addEventListener("keydown", function(event)
+		{
+			if(event.keyCode === 9) // TAB characterSet
+			{			
+				//let focus_query = 'button, [href], input, select, textarea, [tabIndex]:not([tabIndex="-1"])';
+				let focus_query = 'button, [href], input, select, textarea';
+				var focusable = Array(...control.content.querySelectorAll(focus_query));
+				if(control.hasOwnProperty("div_buttons"))
+				{
+					focusable.push(...control.div_buttons.querySelectorAll(focus_query));
+				}			
+				
+				if(focusable.length === 0)
+				{
+					(document.activeElement as any)?.blur();
+				}
```

Comment:
Are you sure `.blur()` is the right thing to do here? When there are no focusable elements in the dialog (which is already very unlikely), why not focus the container?

Reply:
Now I changed it to behave similar to when the dialog is opened. The `blur` is now not needed anymore, because the content `div` of the dialog is made explicitly focusable by Javascript with `tabindex="-1"`, in this case `.focus()` should not fail. Fix: The `event.preventDefault()` was missing. I have tested this by temporarily removing the buttons from the `[F1]` popup, that dialog would have no focusable elements.